python/dgl/data/explain_syn.py

```diff
@@ -0,0 +1,757 @@
+        Default: False
+    verbose : bool, optional
+        Whether to print progress information. Default: True
+    transform : callable, optional
+        A transform that takes in a :class:`~dgl.DGLGraph` object and returns
+        a transformed version. The :class:`~dgl.DGLGraph` object will be
+        transformed before every access. Default: None
+
+    Attributes
+    ----------
+    num_classes : int
+        Number of node classes
+
+    Examples
+    --------
+
+    >>> from dgl.data import TreeCycles
+    >>> dataset = TreeCycles()
+    >>> dataset.num_classes
+    2
+    >>> g = dataset[0]
+    >>> label = g.ndata['label']
+    >>> feat = g.ndata['feat']
+    """
+    def __init__(self,
+                 tree_height=8,
+                 num_motifs=60,
+                 perturb_ratio=0.01,
+                 seed=None,
+                 raw_dir=None,
+                 force_reload=False,
+                 verbose=True,
+                 transform=None):
+        self.tree_height = tree_height
+        self.num_motifs = num_motifs
+        self.perturb_ratio = perturb_ratio
+        self.seed = seed
+        super(TreeCycles, self).__init__(name='TREE-CYCLES',
+                                         url=None,
+                                         raw_dir=raw_dir,
+                                         force_reload=force_reload,
+                                         verbose=verbose,
+                                         transform=transform)
+
+    def process(self):
+        if self.seed is not None:
+            np.random.seed(self.seed)
+
+        g = nx.balanced_tree(r=2, h=self.tree_height)
+        edges = list(g.edges())
+        src, dst = map(list, zip(*edges))
+        n = nx.number_of_nodes(g)
+
+        # Nodes in the base tree graph belong to class 0
+        node_labels = [0] * n
+        # The motifs will be evenly attached to the nodes in the base graph.
+        spacing = math.floor(n / self.num_motifs)
+
+        cycle_size = 6
+        for motif_id in range(self.num_motifs):
+            # Construct a six-node cycle
+            motif_edges = [(n + i, n + i + 1) for i in range(5)]
+            motif_edges.append((n + 5, n))
+            motif_src, motif_dst = map(list, zip(*motif_edges))
+            src.extend(motif_src)
+            dst.extend(motif_dst)
+
+            # Nodes in cycles belong to class 1
+            node_labels.extend([1] * cycle_size)
+
+            # Attach the motif to the base tree graph
+            anchor = int(motif_id * spacing)
+            src.append(n)
+            dst.append(anchor)
+
+            if np.random.random() > 0.5:
+                a = np.random.randint(1, 4)
+                b = np.random.randint(1, 4)
+                src.append(n + a)
+                dst.append(anchor + b)
+
+            n += cycle_size
+
+        g = graph((src, dst), num_nodes=n)
+
+        # Perturb the graph by adding non-self-loop random edges
```

Comment:
Do we have a utility function to perturb a graph?

Reply:
One possible solution is to extend [`AddEdge`](https://docs.dgl.ai/generated/dgl.transforms.AddEdge.html#dgl.transforms.AddEdge) transform.